docs/in-depth.md

```diff
@@ -0,0 +1,220 @@
+
+In order to select a single *slot* in our group, we might write something simple like:
+
+```
+function chooseOperator(operators) {
+  const operatorNames = Object.keys(operators)
+
+  let totalWeight = 0
+  for (let i = 0; i < operatorNames.length; i++) {
+    const operatorName = operatorNames[i]
+    totalWeight += operators[operatorName]
+  }
+
+  let randomValue = Math.floor(Math.random() * totalWeight)
+
+  let chosenOperator = undefined
+  for (let i = 0; i < operatorNames.length; i++) {
+    const operatorName = operatorNames[i]
+    const operatorStake = operators[operatorName]
+    if (operatorStake > randomValue) {
+      chosenOperator = operatorName
+      break;
+    } else {
+      randomValue -= operatorStake
+    }
+  }
+  return chosenOperator
+}
+chooseOperator({"Alice": 20, "Bob": 30, "Carol": 40, "David": 50})
+```
+
+In english, we figure out the total weight of the pool, and then generate a
+random number in `[0, totalWeight)`. Then, we assign a sort of cumulative
+distribution to each of the operators: Alice is picked if that random number is
+in `[0, 20)`, Bob is picked if the random number is in `[20, 50)`, Carol is
+picked if the random number is in `[50, 90)`, and David is picked if the random
+number is in `[90, 140)`.
+
+Say our random number was 64. We know that's in Carol's range, but how do we
+determine that? First, we check if 64 > Alice's 20, and it isn't, so we
+subtract 20 from 64 and move to Bob. Bob's 30 still isn't greater than the
+remaining 44, so we subtract 30 and move to Carol. Carol's 40 is greater than
+the remaining 14 so we're finished.
+
+If we need to generate `numOperators` amount of operators, then we just loop:
+
+```
+function chooseOperators(operators, numOperators) {
+let group = []
+  for (let i = 0; i < numOperators; i++) {
+    group.push(chooseOperator(operators))
+  }
+  return group
+}
+chooseOperators({"Alice": 20, "Bob": 30, "Carol": 40, "David": 50}, 3)
+```
+
+This just works! In applications where performance doesn't matter and we don't
+care about the cost of reading or writing data to memory, this is probably good
+enough. On ethereum, looping through thousands of operators a hundred times to
+get a 100-person group is *extremely* expensive, so we need to optimize.
+
+### Binary Trees
+
+Say that we have 1000 operators, each of which has 1 stake. In the average
+case, we need to iterate through 500 operators, and in the worst case, we need
+to iterate through all 1000 of them. We can do better!
+
+If we store the operators in a binary tree where each branch is the sum of the
+weight of the branches below it, and the leaves are the individual operators
+with their individual weights, then we have a bigger data structure that is
+*much* faster to search.
+
+We can represent 2 operators with a 1-layer tree, 4 operators with a 2-layer
+tree, 8 operators with a 3-layer, and `2^n` operators with a `n`-layer tree. In
+order to represent our 1000 operators, we need 10 layers.
```

Comment:
This assumes that we don't count the root as a layer, right? Intuitively, I would say that the tree from the example in line 106 (with 4 operators) has 3 layers and that would mean that we need `n`-layer tree to represent `2^(n-1)` operators. I've did a quick googling and it looks that in graph theory, the root level is usually counted as a layer (for example, see [here - graphs on p.10 ](http://cs.brown.edu/people/rtamassi/gdhandbook/chapters/hierarchical.pdf)).
I would suggest either counting the root as a layer or changing text to something like "We can represent 2 operators with a 1-layer tree (not counting root as a layer)..."

Reply:
https://github.com/keep-network/sortition-pools/pull/157/commits/5af75c3efa8213f31d80cba70a19266a8e9fe347